server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v64/MakeComponentsPrivateBaseOnPermissions.java

```diff
@@ -0,0 +1,269 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.platform.db.migration.version.v64;
+
+import java.sql.SQLException;
+import org.sonar.db.Database;
+import org.sonar.server.platform.db.migration.step.DataChange;
+import org.sonar.server.platform.db.migration.step.MassUpdate;
+import org.sonar.server.platform.db.migration.step.Select;
+import org.sonar.server.platform.db.migration.step.SqlStatement;
+
+/**
+ * This DB migration assumes the whole PROJECTS table contains only rows with value false
+ * (set by {@link PopulateColumnProjectsPrivate}) and performs the following:
+ * <ul>
+ *   <li>set private=true for any tree of component which root has neither user nor codeviewer permission for group AnyOne
+ *       but defines at least one permission directly to a group or user</li>
+ *   <li>removes any permission to group AnyOne for root components which are made private</li>
+ *   <li>ensures any group or user with direct permission to a private root component also has both permissions
+ *       user and codeviewer</li>
+ *   <li>deletes any permission user or codeviewer for root components which stays public</li>
+ * </ul>
+ */
+public class MakeComponentsPrivateBaseOnPermissions extends DataChange {
+
+  private static final String SCOPE_PROJECT = "PRJ";
+  private static final String QUALIFIER_PROJECT = "TRK";
+  private static final String QUALIFIER_VIEW = "VW";
+  private static final String PERMISSION_USER = "user";
+  private static final String PERMISSION_CODEVIEWER = "codeviewer";
+
+  public MakeComponentsPrivateBaseOnPermissions(Database db) {
+    super(db);
+  }
+
+  @Override
+  protected void execute(Context context) throws SQLException {
+    makePrivateComponent(context);
+    cleanPermissionsOfPublicComponents(context);
+    insertUserPermissionOfPrivateRootComponent(context, PERMISSION_USER);
+    insertUserPermissionOfPrivateRootComponent(context, PERMISSION_CODEVIEWER);
+    insertGroupPermissionOfPrivateRootComponent(context, PERMISSION_USER);
+    insertGroupPermissionOfPrivateRootComponent(context, PERMISSION_CODEVIEWER);
+  }
+
+  private static void makePrivateComponent(Context context) throws SQLException {
+    MassUpdate massUpdate = context.prepareMassUpdate();
+    massUpdate.select("select uuid, id from projects p where " +
+      " p.scope = ?" +
+      " and p.qualifier in (?, ?)" +
+      " and p.private = ?" +
+      " and not exists (" +
+      "   select" +
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "   select" 3 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1192)